tiledb/sm/rest/rest_client.cc

```diff
@@ -260,50 +261,175 @@ Status RestClient::post_query_submit(
+    // the offset to point to the start of the serialized query.
+    scratch->advance_offset(8);
+
+    // We can only deserialize the query if it is 8-byte aligned. If the
+    // offset is 8-byte aligned, we can deserialize the query in-place.
+    // Otherwise, we must make a copy to an auxiliary buffer.
+    if (scratch->offset() % 8 != 0) {
+      // Copy the entire serialized buffer to a newly allocated, 8-byte
+      // aligned auxiliary buffer.
+      Buffer aux;
+      st = aux.write(scratch->cur_data(), query_size);
+      if (!st.ok()) {
+        scratch->set_offset(scratch->offset() - 8);
+        scratch->set_size(scratch->offset());
+        return std::max(bytes_processed, 0L);
+      }
+
+      // Deserialize the buffer and store it in 'copy_state'. If
+      // the user buffers are too small to accomodate the attribute
+      // data when deserializing read queries, this will return an
+      // error status.
+      st = serialization::query_deserialize(
+          aux, serialization_type_, true, copy_state, query);
+      if (!st.ok()) {
+        scratch->set_offset(scratch->offset() - 8);
+        scratch->set_size(scratch->offset());
+        return std::max(bytes_processed, 0L);
+      }
+    } else {
+      // Deserialize the buffer and store it in 'copy_state'. If
+      // the user buffers are too small to accomodate the attribute
+      // data when deserializing read queries, this will return an
+      // error status.
+      st = serialization::query_deserialize(
+          *scratch, serialization_type_, true, copy_state, query);
+      if (!st.ok()) {
+        scratch->set_offset(scratch->offset() - 8);
+        scratch->set_size(scratch->offset());
+        return std::max(bytes_processed, 0L);
+      }
     }
 
-    // The returned data may not be 8-byte aligned. Allocate a new buffer and
-    // copy the serialized buffer into it for deserialization.
-    Buffer aux;
-    RETURN_NOT_OK(aux.write(
-        static_cast<char*>(returned_data.data()) + offset, query_size))
+    scratch->advance_offset(query_size);
+    bytes_processed += (query_size + 8);
+  }
 
-    // Deserialize data returned. If the user buffers are too small to
-    // accomodate the attribute data when deserializing read queries, this will
-    // return an error status.
-    RETURN_NOT_OK(serialization::query_deserialize(
-        aux, serialization_type_, true, copy_state, query));
+  // If there are unprocessed bytes left in the scratch space, copy them
+  // to the beginning of 'scratch'. The intent is to reduce memory
+  // consumption by overwriting the serialized query objects that we
+  // have already processed.
+  const uint64_t length = scratch->size() - scratch->offset();
+  if (scratch->offset() != 0) {
+    const uint64_t offset = scratch->offset();
+    scratch->set_offset(0);
+
+    if (length > offset) {
+      st = scratch->write(scratch->data(offset), length);
```

Comment:
Couple of things here: 
* The if branches should be swapped
* Use a local `Buffer` instead of malloc/free
* Add a brief comment about memcpy to explain the temp buffer dance

Reply:
- D'oh
- Done
- Done